cnd/src/libp2p_bam/substream/in.rs

```diff
@@ -0,0 +1,169 @@
+use crate::libp2p_bam::{
+    handler::{AutomaticallyGeneratedErrorResponse, InnerEvent, PendingIncomingRequest},
+    protocol::BamStream,
+    substream::{Advance, Advanced, CloseStream},
+};
+use bam::{
+    json::{
+        Frame, FrameType, Header, Response, UnknownMandatoryHeaders, UnvalidatedIncomingRequest,
+    },
+    IntoFrame, Status,
+};
+use futures::sync::oneshot;
+use libp2p::core::protocols_handler::ProtocolsHandlerEvent;
+use std::collections::{HashMap, HashSet};
+use tokio::prelude::*;
+
+#[derive(strum_macros::Display)]
+#[allow(missing_debug_implementations)]
+/// States of an incoming substream i.e. from peer node to us.
+pub enum State<TSubstream> {
```

Comment:
I don't understand what `WaitingUser` & `WaitingSend` are.
Can you explain what this the cycle for this `State` enum?

Reply:
Looks like we got a bit too excited doing the last patch of this PR (removing comments on enum variants).  I have ammended (and force pushed) the last patch to re-add the variant comments.  Does this answer your question enough @D4nte?  If not then the code is not clear enough IMO or the comments need improving some more.  Thanks

Reply:
> If not then the code is not clear enough IMO or the comments need improving some more.

Not sure how much up-front knowledge about libp2p and the messaging should be assumed and how much we want to explain in comments in the code?

Reply:
@D4nte have you had a look through the commit messages yet? They also contain useful information :)

Reply:
> > If not then the code is not clear enough IMO or the comments need improving some more.
> 
> Not sure how much up-front knowledge about libp2p and the messaging should be assumed and how much we want to explain in comments in the code?

IMHO, Enough so that the reviewer understand what he approves.